### PR TITLE
cmd/tailscaled: change Windows service shutdown and add optional even…

### PR DESCRIPTION
**Pull request by dblohm7**
…t logging

Once a stop request is received and the service updates its status to `svc.StopPending`,
it should continue running *until the shutdown sequence is complete*, and then
return out of `(*ipnService).Execute`, which automatically sends a `svc.Stopped`
notification to Windows.

To make this happen, I changed the loop so that it runs until `doneCh` is
closed, and then returns. I also removed a spurious `svc.StopPending` notification
that the Windows Service Control Manager might be interpreting as a request for
more time to shut down.

Finally, I added some optional logging that sends a record of service notifications
to the Windows event log, allowing us to more easily correlate with any Service
Control Manager errors that are sent to the same log.

Signed-off-by: Aaron Klotz <aaron@tailscale.com>

### CODE REVIEW

File: cmd/tailscaled/tailscaled_windows.go [hunk 1]
```diff
@@ -60,12 +61,31 @@ func isWindowsService() bool {
 	return v
 }
 
+// syslogf is a logger function that writes to the Windows event log (ie, the
+// one that you see in the Windows Event Viewer). tailscaled may optionally
+// generate diagnostic messages in the same event timeline as the Windows
+// Service Control Manager to assist with diagnosing issues with tailscaled's
+// lifetime (such as slow shutdowns).
+var syslogf logger.Logf = func(string, ...any) {}
```
Comment (bradfitz) [inline cmd/tailscaled/tailscaled_windows.go:69]: `= logger.Discard`

**Comment by dblohm7** [comment]
I'm not sure this fix is as cut and dry as I originally thought it would be, but I think the event log entries would be pretty useful to have.

File: cmd/tailscaled/tailscaled_windows.go [hunk 2]
```diff
@@ -98,26 +121,29 @@ func (service *ipnService) Execute(args []string, r <-chan svc.ChangeRequest, ch
...
 		select {
 		case <-doneCh:
+			return false, windows.NO_ERROR
 		case cmd := <-r:
 			log.Printf("Got Windows Service event: %v", cmdName(cmd.Cmd))
 			switch cmd.Cmd {
 			case svc.Stop:
+				changes <- svc.Status{State: svc.StopPending}
+				syslogf("Service stop pending")
 				cancel()
```
Comment (bradfitz) [inline cmd/tailscaled/tailscaled_windows.go:136]: cancel() // so BabysitProc will kill child process

(I had to remind myself how that worked)

**Review by bradfitz — APPROVED** [review]
The depaware changes seem surprising?

**Comment by dblohm7** [comment]
> The depaware changes seem surprising?

I agree, yet CI fails without them.

**Comment by bradfitz** [comment]
> I'm not sure this fix is as cut and dry as I originally thought it would be, but I think the event log entries would be pretty useful to have.

Yeah, seems fine to keep that debug in.

**Comment by bradfitz** [comment]
> I agree, yet CI fails without them.

Is this a Windows thing perhaps? Do you normally develop with a different primary GOOS?

**Comment by dblohm7** [comment]
I build using WSL. I set `GOOS=windows` when building binaries, but I ran `make depaware` without any extra env overrides.

**Comment by bradfitz** [comment]
I only get:

```patch
bradfitz@tsdev:~/src/tailscale.com$ git di
diff --git a/cmd/tailscaled/depaware.txt b/cmd/tailscaled/depaware.txt
index 998536d0..97c4fd28 100644
--- a/cmd/tailscaled/depaware.txt
+++ b/cmd/tailscaled/depaware.txt
@@ -323,6 +323,7 @@ tailscale.com/cmd/tailscaled dependencies: (generated by github.com/tailscale/de
    W    golang.org/x/sys/windows                                     from github.com/go-ole/go-ole+
    W    golang.org/x/sys/windows/registry                            from golang.zx2c4.com/wireguard/windows/tunnel/winipcfg+
    W    golang.org/x/sys/windows/svc                                 from golang.org/x/sys/windows/svc/mgr+
+   W    golang.org/x/sys/windows/svc/eventlog                        from tailscale.com/cmd/tailscaled
    W    golang.org/x/sys/windows/svc/mgr                             from tailscale.com/cmd/tailscaled
         golang.org/x/term                                            from tailscale.com/logpolicy
         golang.org/x/text/secure/bidirule                            from golang.org/x/net/idna
```

Maybe you had an old git parent? The CI runs against a merge commit of your commit and HEAD.

**Comment by bradfitz** [comment]
I don't know how your `--update` added so many lines but I think the problem is `--check` only checks for missing edges, not extra edges. I'll fix the latter.

I did a force push with my `--update` results to see if CI's also happy with that.

**Pull request merged**